File: src/java/magma/collect/stream/SingleHead.java
package magma.collect.stream;

import magma.option.None;
import magma.option.Option;
import magma.option.Some;

public class SingleHead<T> implements Head<T> {
    private final T value;
    private boolean retrieved = false;

    public SingleHead(T value) {
        this.value = value;
    }

    @Override
    public Option<T> next() {
        if (retrieved) return new None<>();

        retrieved = true;
        return new Some<>(value);
    }
}

File: src/windows/magma/api/collect/head/SingleHead.c
#include "./SingleHead.h"
export class SingleHead<T> implements Head<T> {
	element: T;
	retrieved: Bool;
}

constructor (element: T) {
	this.element = element;
	this.retrieved = false;
}
next(): Option<T> {
	if (this.retrieved) {
		return new None<T>();
	}
	this.retrieved = true;
	return new Some<T>(this.element);
}